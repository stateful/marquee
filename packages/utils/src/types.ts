import type { IconProp } from '@fortawesome/fontawesome-svg-core'
import type { TelemetryEventProperties } from '@vscode/extension-telemetry'
import type { Webview } from 'vscode'

export type ContextProperties<T> = {
  [t in keyof T]: T[t]
} & {
  [t in keyof T & string as `set${Capitalize<t>}`]: (val: T[t]) => void
}

interface VSCodeWebview extends Webview {
  getState: () => any
  setState: (param: any) => void
}

export interface ThirdPartyWidgetOptions {
  name: string
  icon: IconProp
  label: string
  tags: string[]
  description: string
}

export interface MarqueeInterface {
  defineWidget: (
    widgetOptions: ThirdPartyWidgetOptions,
    constructor: CustomElementConstructor,
    options?: ElementDefinitionOptions
  ) => void
}

export interface MarqueeWindow<State = any, Configuration = any> extends Window {
  vscode: VSCodeWebview
  acquireVsCodeApi: () => Webview
  activeWorkspace: Workspace | null
  marqueeUserProps: string
  marqueeExtension: MarqueeInterface
  marqueeBackendBaseUrl: string
  marqueeBackendGeoUrl: string
  marqueeBackendFwdGeoUrl: string
  marqueeStateConfiguration: Record<string, { state: State, configuration: Configuration }>
  marqueeThirdPartyWidgets: number
}

export interface MarqueeEvents {
  openSettings: never
  removeWidget: string
  updateWidgetDisplay: Record<string, boolean>
  resetMarquee?: boolean
  telemetryEvent: {
    eventName: string
    properties?: TelemetryEventProperties
  }
  openCloudSyncFeatureInterest: boolean
}

export enum WorkspaceType {
  WORKSPACE = 'workspace',
  FOLDER = 'folder',
  NONE = 'none',
}

export interface Workspace {
  id: string
  name: string
  path: string
  type: WorkspaceType
}

export type RGBA = {
  r: number;
  g: number;
  b: number;
  a?: number;
}

export interface Configuration {
  background: string
  name: string
}

export interface State extends Pick<ProjectItem, 'commit' | 'branch' | 'gitUri'> {
  globalScope: boolean
}

export interface GuiState {
  resetApp: boolean
}

export interface Context extends ContextProperties<State & Configuration & GuiState> {
  themeColor: RGBA
}

export interface ProjectItem {
  commit?: string
  branch?: string
  gitUri?: string
  path?: string
  origin?: string
  body: string
  id: string
  workspaceId: string | null
}

export type ProjectItemTypes = 'todos' | 'snippets' | 'notes'

export interface GitRemote {
  name: string;
  url: string;
}

export interface StateManager<State> {
  get (): State
  get (property: keyof State): State[keyof State]
  set (property: keyof State, value: State[keyof State]): Promise<void>
}
