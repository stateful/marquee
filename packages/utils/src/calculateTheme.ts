import { createTheme, ThemeOptions } from '@mui/material/styles'
import React from 'react'

export default () => {
  const fontSize = parseInt(
    window.getComputedStyle(document.documentElement)
      .getPropertyValue('--vscode-font-size')
      .replace('px', '')
  )

  const vscodeStyles = document.documentElement.style.cssText.split(';')
  const vsCodeStyleMap: Record<string, string> = {}

  vscodeStyles.forEach((style) => {
    if (style.indexOf('--vscode') !== -1) {
      const pair = style.replace(/\s/g, '').split(':')
      vsCodeStyleMap[pair[0]] = pair[1]
    }
  })

  const newTheme: ThemeOptions = {
    shape: {
      borderRadius: 0,
    },
    components: {
      MuiSvgIcon: {
        defaultProps: {
          htmlColor: vsCodeStyleMap['--vscode-icon-foreground']
        }
      },
      MuiChip: {
        styleOverrides: {
          root: {
            borderRadius: '0px',
          }
        }
      },
      MuiButton: {
        styleOverrides: {
          root: {
            color: 'var(--vscode-button-foreground)',
            backgroundColor: 'var(--vscode-button-background)',
            borderColor: 'var(--vscode-editorGhostText-foreground)',
            '&:hover': {
              backgroundColor: 'var(--vscode-button-hoverBackground)',
              borderColor: 'var(--vscode-tab-inactiveForeground)'
            }
          },
          outlined: {
            backgroundColor: 'transparent',
            color: 'var(--vscode-button-background)',
          }
        }
      },
      MuiListItem: {
        styleOverrides: {
          root: {
            '&:hover': {
              backgroundColor: 'var(--vscode-button-secondaryHoverBackground)',
              color: 'var(--vscode-button-secondaryForeground)'
            }
          }
        }
      },
      MuiButtonGroup: {
        styleOverrides: {
          grouped: {
            '&:not(:last-of-type)': {
              borderRight: '1px solid rgba(255, 255, 255, 0.23)'
            }
          }
        }
      },
      MuiDialog: {
        styleOverrides: {
          paper: {
            backgroundImage: 'none'
          }
        }
      },
      MuiPopover: {
        styleOverrides: {
          paper: {
            backgroundImage: 'none'
          }
        }
      },
      MuiTabs: {
        styleOverrides: {
          indicator: {
            left: 0
          }
        }
      }
    },
    palette: {
      mode: document.body.classList[0] === 'vscode-light' ? 'light' : 'dark',
      divider: vsCodeStyleMap['--vscode-foreground'],
      background: {
        default: vsCodeStyleMap['--vscode-sideBar-background'],
        paper: vsCodeStyleMap['--vscode-sideBar-background'],
      },
      text: {
        primary: vsCodeStyleMap['--vscode-foreground'],
        secondary: vsCodeStyleMap['--vscode-foreground'],
      },
      primary: {
        main: vsCodeStyleMap['--vscode-button-background'] || '#000000',
        dark: vsCodeStyleMap['--vscode-button-hoverBackground'],
        contrastText: vsCodeStyleMap['--vscode-button-foreground'],
      },
      secondary: {
        main: vsCodeStyleMap['--vscode-editorMarkerNavigationError-background'],
        contrastText: vsCodeStyleMap['--vscode-button-foreground'],
      },
      action: {
        active: vsCodeStyleMap['--vscode-foreground'],
        selected: vsCodeStyleMap['--vscode-editor-selectionBackground'],
        hover: vsCodeStyleMap['--vscode-editor-hoverHighlightBackground'],
        disabled: vsCodeStyleMap['--vscode-editor-inactiveSelectionBackground'],
        disabledBackground:
          vsCodeStyleMap['--vscode-editor-inactiveSelectionBackground'],
      },
    },
    typography: {
      fontFamily: vsCodeStyleMap['--vscode-font-family'],
      fontWeightLight: vsCodeStyleMap['--vscode-font-weight'] as React.CSSProperties['fontWeight'],
      fontWeightRegular: vsCodeStyleMap['--vscode-font-weight'] as React.CSSProperties['fontWeight'],
      fontWeightMedium: vsCodeStyleMap['--vscode-font-weight'] as React.CSSProperties['fontWeight'],
      fontWeightBold: vsCodeStyleMap['--vscode-font-weight'] as React.CSSProperties['fontWeight'],
      fontSize: fontSize,
      button: {
        textTransform: 'none',
      },
    }
  }

  return createTheme(newTheme)
}
