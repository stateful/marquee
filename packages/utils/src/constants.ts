import type { State, Configuration } from './types'

declare const PACKAGE_JSON: string
export const pkg: any = PACKAGE_JSON
export const EXTENSION_ID = 'stateful.marquee'
export const defaultConfigurations = pkg.contributes.configuration[0].properties
export const DEPRECATED_GLOBAL_STORE_KEY = 'persistence'

export const DEFAULT_STATE: State = {
  globalScope: false
}

export const DEFAULT_CONFIGURATION: Configuration = {
  background: defaultConfigurations['marquee.configuration.background'].default,
  name: defaultConfigurations['marquee.configuration.name'].default
}
