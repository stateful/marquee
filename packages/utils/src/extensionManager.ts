import vscode from 'vscode';
import crypto from 'crypto';
import { Client } from 'tangle';
import { EventEmitter } from 'events';

import { Workspace, WorkspaceType } from './types';

const config = vscode.workspace.getConfiguration('marquee');

export default class ExtensionManager<State, Configuration> extends EventEmitter implements vscode.Disposable {
  private _tangle?: Client<State & Configuration>;
  protected _state: State;
  protected _configuration: Configuration;

  constructor (
    private _context: vscode.ExtensionContext,
    private _channel: vscode.OutputChannel,
    private _key: string,
    private _defaultConfiguration: Configuration,
    private _defaultState: State
  ) {
    super();
    this._state = this._context.globalState.get<State>(this._key) || this._defaultState;
    this._configuration = config.get<Configuration>(this._key) || this._defaultConfiguration;
  }

  updateConfiguration <T extends keyof Configuration>(prop: T, val: Configuration[T]) {
    this._channel.appendLine(`Update configuration "${prop}": ${val}`);
    this._configuration[prop] = val;
    config.update(`${this._key}.${prop}`, val, vscode.ConfigurationTarget.Global);
  }

  updateState <T extends keyof State>(prop: T, val: State[T]) {
    this._channel.appendLine(`Update state "${prop}": ${val}`);
    this._state[prop] = val;
    this._context.globalState.update(this._key, this._state);
  }

  setBroadcaster (tangle: Client<State & Configuration>) {
    this._tangle = tangle;

    /**
     * listen on configuration changes
     */
    for (const configProp of Object.keys(this._defaultConfiguration)) {
      const c = configProp as keyof Configuration;
      this._tangle.listen(c, (val) => this.updateConfiguration(c, val));
    }

    /**
     * listen on state changes
     */
    for (const stateProp of Object.keys(this._defaultState)) {
      const s = stateProp as keyof State;
      this._tangle.listen(s, (val) => this.updateState(s, val));
    }
  }

  protected getActiveWorkspace(): Workspace | null {
    const wsp = vscode.workspace;
    let name = wsp.name || "";
    let path = "";
    let type = WorkspaceType.NONE;

    if (wsp.workspaceFile) {
      type = WorkspaceType.WORKSPACE;
      path = wsp.workspaceFile.path;
    } else if (wsp.workspaceFolders) {
      type = WorkspaceType.FOLDER;
      path =
        wsp.workspaceFolders.length > 0 ? wsp.workspaceFolders[0].uri.path : "";
    }

    if (type && path) {
      const shasum = crypto.createHash("sha1");
      const id = shasum.update(path, "utf8").digest("hex");
      const nws: Workspace = { id, name, type, path };

      return nws;
    }

    return null;
  }

  dispose() {
    delete this._tangle;
  }
}
