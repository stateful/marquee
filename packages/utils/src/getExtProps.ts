import os from "os";
import fs from 'fs';
import path from 'path';
import vscode from "vscode";

const EXTENSION_ID = "stateful.marquee";
const TELEMETRY_CONFIG_ID = "telemetry";
const TELEMETRY_CONFIG_ENABLED_ID = "enableTelemetry";

export function getMarqueeVersion (context: vscode.ExtensionContext) {
  const extensionPath = path.join(context.extensionPath, "package.json");
  const packageFile = fs.readFileSync(extensionPath, "utf-8");
  const packageJson = JSON.parse(packageFile);
  return packageJson.version;
}

export default function getExtProps(context: vscode.ExtensionContext): { [key: string]: string } {
  const extProps = Object.create(null);

  const config = vscode.workspace.getConfiguration(TELEMETRY_CONFIG_ID);
  if (!config.get<boolean>(TELEMETRY_CONFIG_ENABLED_ID, true)) {
    return extProps;
  }

  extProps["os"] = os.platform();
  extProps["platformversion"] = (os.release() || "").replace(
    /^(\d+)(\.\d+)?(\.\d+)?(.*)/,
    "$1$2$3"
  );
  extProps["extname"] = EXTENSION_ID;
  extProps["extversion"] = getMarqueeVersion(context);
  if (vscode && vscode.env) {
    extProps["vscodemachineid"] = vscode.env.machineId;
    extProps["vscodesessionid"] = vscode.env.sessionId;
    extProps["vscodeversion"] = vscode.version;

    switch (vscode.env.uiKind) {
      case vscode.UIKind.Web:
        extProps["uikind"] = "web";
        break;
      case vscode.UIKind.Desktop:
        extProps["uikind"] = "desktop";
        break;
      default:
        extProps["uikind"] = "unknown";
    }
  }
  return extProps;
}
