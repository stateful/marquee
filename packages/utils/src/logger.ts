import vscode from 'vscode'
import { getExtensionLogger, IVSCodeExtLogger, IChildLogger, LogLevel } from '@vscode-logging/logger'

declare const IS_WEB_BUNDLE: boolean
const DEFAULT_LOG_LEVEL: LogLevel = 'info'

interface MessagePayload {
  label: string
  level: string
  message: string
  time: string
}

export type ChildLogger = IChildLogger

export class Logger {
  private static output?: IVSCodeExtLogger

  static configure (context: vscode.ExtensionContext) {
    const { name, publisher } = context.extension.packageJSON
    const inDevelopment = context.extensionMode === vscode.ExtensionMode.Development

    const channel = vscode.window.createOutputChannel(name.slice(0, 1).toUpperCase() + name.slice(1))
    this.output = getExtensionLogger({
      extName: `${publisher}.${name}`,
      level: this.logLevel, // See LogLevel type in @vscode-logging/types for possible logLevels
      ...(IS_WEB_BUNDLE
        ? {}
        : { logPath: context.logUri.fsPath }
      ),
      logOutputChannel: <vscode.OutputChannel>{
        appendLine: (payload: string) => {
          const msg: MessagePayload = JSON.parse(payload)
          return channel.appendLine(`[${msg.time}] ${msg.level.toUpperCase()} ${msg.label} - ${msg.message}`)
        }
      }, // OutputChannel for the logger
      sourceLocationTracking: false,
      logConsole: inDevelopment // define if messages should be logged to the consol
    })
  }

  private static _logLevel: LogLevel = DEFAULT_LOG_LEVEL
  static get logLevel (): LogLevel {
    return this._logLevel
  }
  static set logLevel (value: LogLevel) {
    this._logLevel = value
    this.output?.changeLevel(value)
  }

  static getChildLogger (label: string): ChildLogger {
    if (!this.output) {
      throw new Error('Can\'t create child logger if main logger wasn\'t initiated')
    }

    return this.output.getChildLogger({ label })
  }

  static info (message: string, ...args: any[]) {
    this.output?.info(message, ...args)
  }

  static warn (message: string, ...args: any[]) {
    this.output?.warn(message, ...args)
  }

  static error (message: string, ...args: any[]) {
    this.output?.error(message, ...args)
  }

  static fatal (message: string, ...args: any[]) {
    this.output?.fatal(message, ...args)
  }

  static debug (message: string, ...args: any[]) {
    this.output?.debug(message, ...args)
  }

  static trace (message: string, ...args: any[]) {
    this.output?.trace(message, ...args)
  }
}
