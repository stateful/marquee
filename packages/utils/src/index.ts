/* istanbul ignore file */

import { useState, useEffect } from 'react'
import Channel from 'tangle/webviews'

import type { Client } from 'tangle'

import calculateTheme from './calculateTheme'

import BetterComplete from './components/BetterComplete'
import NetworkError from './components/NetworkError'
import DoubleClickHelper from './components/DoubleClickHelper'
import ProjectItemLink from './components/ProjectItemLink'
import { jumpTo } from './components/utils'

import GlobalContext, { GlobalProvider } from './contexts/Global'
import type { MarqueeWindow, ContextProperties } from './types'

const defaultChannel = 'vscode.marquee'
const theme = calculateTheme()

declare const window: MarqueeWindow

const tangleChannels = new Map<string, any>()
const getEventListener = <T>(channel = defaultChannel) => {
  if (!tangleChannels.has(channel)) {
    const ch = new Channel<T>(channel)
    const eventListener = ch.attach(window.vscode as any)
    tangleChannels.set(channel, eventListener)
    return eventListener
  }
  return tangleChannels.get(channel)! as Client<T>
}

type Entries<T> = {
  [K in keyof T]: [K, T[K]]
}[keyof T][]

/**
 * Helper method to connect a widget context with its configuration and
 * application state of the extension host
 *
 * @param defaults combind default values for configuration and state
 * @param tangle   tangle client to broadcast information
 * @returns object containing state values and its setter methods
 */
function connect<T, Events = {}> (defaults: T, tangle: Client<T & Events>): ContextProperties<T> {
  const prevState: any = window.vscode.getState() || {}
  const contextValues: Partial<ContextProperties<T>> = {}
  for (const [prop, defaultVal] of Object.entries(defaults as any) as Entries<ContextProperties<T>>) {
    /**
     * get default state either from:
     * - the webview state (when widget was removed and re-loaded)
     * - the actual default value
     */
    const defaultState = typeof prevState[prop] === 'undefined'
      ? defaultVal
      : prevState[prop]

    const [propState, setPropState] = useState<typeof defaultVal>(defaultState)
    contextValues[prop as keyof T] = propState as ContextProperties<T>[keyof T]

    /**
     * define a custom setter method for a state that:
     * - updates the state
     * - broadcasts it to the extension host for sync
     * - add it to the webview state so if a webview is being revived it
     *   receives its former configurations
     */
    const firstLetter = (prop as string).slice(0, 1).toUpperCase()
    const restLetters = (prop as string).slice(1)
    const setProp = `set${firstLetter}${restLetters}` as `set${Capitalize<keyof T & string>}`
    contextValues[setProp] = ((val: any) => {
      setPropState(val)
      tangle.broadcast({ [prop]: val } as T & Events)
      window.vscode.setState({
        ...(window.vscode.getState() || {}),
        [prop]: val
      })
    }) as any

    /**
     * listen to updates from the extension host and apply it to the property
     * state and webview state
     */
    useEffect(() => {
      const subs = tangle.listen(prop as keyof T, (val: any) => {
        setPropState(val)
        window.vscode.setState({
          ...(window.vscode.getState() || {}),
          [prop]: val
        })
      })

      return () => { subs?.unsubscribe() }
    }, [])
  }

  return contextValues as ContextProperties<T>
}

export {
  jumpTo,
  getEventListener,
  theme,
  connect,

  // components
  BetterComplete,
  NetworkError,
  DoubleClickHelper,
  ProjectItemLink,

  // contexts
  GlobalContext,
  GlobalProvider
}
export default theme
export * from './types'
