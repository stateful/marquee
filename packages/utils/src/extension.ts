import vscode from 'vscode';
import crypto from 'crypto';
import pick from 'lodash.pick';
import { Client } from 'tangle';
import { EventEmitter } from 'events';

import getExtProps from './getExtProps';
import { DEFAULT_CONFIGURATION, DEFAULT_STATE } from './constants';
import { WorkspaceType } from './types';
import type { Configuration, State, Workspace } from './types';

const DEPRECATED_GLOBAL_STORE_KEY = 'persistence';
const CONFIGURATION_TARGET = vscode.ConfigurationTarget.Global;

export default class ExtensionManager<State, Configuration> extends EventEmitter implements vscode.Disposable {
  protected _tangle?: Client<State & Configuration>;
  protected _state: State;
  protected _configuration: Configuration;
  protected _disposables: vscode.Disposable[] = [
    vscode.workspace.onDidChangeConfiguration(this._onConfigChange.bind(this))
  ];
  protected _subscriptions: { unsubscribe: Function }[] = [];
  private _stopListenOnChangeEvents = false;

  constructor (
    protected _context: vscode.ExtensionContext,
    protected _channel: vscode.OutputChannel,
    private _key: string,
    private _defaultConfiguration: Configuration,
    private _defaultState: State
  ) {
    super();
    const config = vscode.workspace.getConfiguration('marquee');

    const oldGlobalStore = this._context.globalState.get<object>(DEPRECATED_GLOBAL_STORE_KEY, {});
    this._state = {
      ...this._defaultState,
      ...pick(oldGlobalStore, Object.keys(this._defaultState)),
      ...this._context.globalState.get<State>(this._key)
    };

    /**
     * preserve state across different machines
     */
    this._context.globalState.setKeysForSync(Object.keys(this._defaultState));

    this._configuration = {
      ...this._defaultConfiguration,
      ...pick<Configuration>(oldGlobalStore as any, Object.keys(this._defaultConfiguration)),
      ...config.get<Configuration>(this._key)
    };
  }

  get state () {
    return this._state;
  }

  get configuration () {
    return this._configuration;
  }

  set stopListenOnChangeEvents (val: boolean) {
    this._stopListenOnChangeEvents = val;
  }

  private _onConfigChange (event: vscode.ConfigurationChangeEvent) {
    /**
     * if the change was triggered by the user through the UI we don't
     * need to act on it because the config change will be already
     * executed
     */
    if (this._stopListenOnChangeEvents) {
      return false;
    }

    for (const configKey of Object.keys(this.configuration)) {
      if (!event.affectsConfiguration(`marquee.${this._key}.${configKey}`)) {
        continue;
      }

      const config = vscode.workspace.getConfiguration('marquee');
      const prop = configKey as keyof Configuration;
      const val = config.get(`${this._key}.${configKey}`) as Configuration[keyof Configuration];
      this._channel.appendLine(`Update configuration via configuration listener "${prop}": ${val}`);
      this._configuration[prop] = val;
      this.broadcast({ [prop]: val } as any);
      break;
    }

    return true;
  }

  protected broadcast (payload: Partial<State & Configuration>) {
    if (!this._tangle) {
      return;
    }

    this._tangle.broadcast(payload as State & Configuration);
  }

  async updateConfiguration <T extends keyof Configuration = keyof Configuration>(prop: T, val: Configuration[T]) {
    const config = vscode.workspace.getConfiguration('marquee');
    this._channel.appendLine(`Update configuration "${prop}": ${val}`);
    this._configuration[prop] = val;
    await config.update(`${this._key}.${prop}`, val, CONFIGURATION_TARGET);
  }

  async updateState <T extends keyof State = keyof State>(prop: T, val: State[T]) {
    this._channel.appendLine(`Update state "${prop}": ${val}`);
    this._state[prop] = val;
    await this._context.globalState.update(this._key, this._state);
    this.emit('stateUpdate', this._state);
  }

  /**
   * clear state and configuration of widget
   */
  public async clear () {
    const config = vscode.workspace.getConfiguration('marquee');
    this._state = { ...this._defaultState };
    await this._context.globalState.update(this._key, this._state);
    await this._context.globalState.update(DEPRECATED_GLOBAL_STORE_KEY, undefined);
    this.emit('stateUpdate', this._state);

    this._configuration = { ...this._defaultConfiguration };
    await Promise.all(
      Object.keys(this._defaultConfiguration).map((key) => (
        config.update(`${this._key}.${key}`, undefined, CONFIGURATION_TARGET)
      ))
    );
    await config.update(this._key, undefined, CONFIGURATION_TARGET);

    /**
     * after we deleted the current configuration, we have to update the
     * object with the default value set in the extension package definition
     * to get its default value from there
     */
    await Promise.all(Object.keys(this._defaultConfiguration).map((key) => {
      type ConfigKey = keyof typeof this._defaultConfiguration;
      const configKey = key as ConfigKey;
      const trueDefault = config.get(`${this._key}.${key}`) as Configuration[ConfigKey] || this._defaultConfiguration[configKey];
      this._configuration[configKey] = trueDefault;
      return config.update(`${this._key}.${key}`, trueDefault, CONFIGURATION_TARGET);
    }));
  }

  /**
   * get current opened workspace
   * @returns workspace object or null if no workspace is opened
   */
  public getActiveWorkspace(): Workspace | null {
    const wsp = vscode.workspace;
    let name = wsp.name || "";
    let path = "";
    let type = WorkspaceType.NONE;

    if (wsp.workspaceFile) {
      type = WorkspaceType.WORKSPACE;
      path = wsp.workspaceFile.path;
    } else if (wsp.workspaceFolders) {
      type = WorkspaceType.FOLDER;
      path =
        wsp.workspaceFolders.length > 0 ? wsp.workspaceFolders[0].uri.path : "";
    }

    if (type && path) {
      const shasum = crypto.createHash("sha1");
      const id = shasum.update(path, "utf8").digest("hex");
      const nws: Workspace = { id, name, type, path };

      return nws;
    }

    return null;
  }

  /**
   * get text selection
   * @param editor vscode.TextEditor
   * @returns selected text
   */
  public getTextSelection(editor: vscode.TextEditor) {
    const textRange = new vscode.Range(
      editor.selection.start.line,
      editor.selection.start.character,
      editor.selection.end.line,
      editor.selection.end.character
    );

    const hier = editor.document.uri.path.split("/");
    const text = editor.document.getText(textRange);
    const name = hier[hier.length - 1];
    const path = `${editor.document.uri.path}:${editor.selection.start.line}`;
    const lang = editor.document.languageId;
    return { text, path, name, lang };
  }

  public setBroadcaster (tangle: Client<State & Configuration>) {
    this._tangle = tangle;

    /**
     * listen on configuration changes
     */
    for (const configProp of Object.keys(this._defaultConfiguration)) {
      const c = configProp as keyof Configuration;
      this._subscriptions.push(this._tangle.listen(c, (val) => this.updateConfiguration(c, val)));
    }

    /**
     * listen on state changes
     */
    for (const stateProp of Object.keys(this._defaultState)) {
      const s = stateProp as keyof State;
      this._subscriptions.push(this._tangle.listen(s, (val) => this.updateState(s, val)));
    }

    return this;
  }

  public generateId (): string {
    const buf = Buffer.alloc(20);
    return crypto.randomFillSync(buf).toString("hex");
  }

  reset () {
    if (this._tangle) {
      this._subscriptions.forEach((s) => s.unsubscribe());
      this._tangle.removeAllListeners();
      delete this._tangle;
    }
  }

  dispose () {
    this._disposables.forEach((d) => d.dispose());
    this.reset();
  }
}

function activate (
  context: vscode.ExtensionContext,
  channel: vscode.OutputChannel
) {
  const stateManager = new ExtensionManager<State, Configuration>(context, channel, 'configuration', DEFAULT_CONFIGURATION, DEFAULT_STATE);

  return {
    marquee: {
      disposable: stateManager,
      defaultState: stateManager.state,
      defaultConfiguration: stateManager.configuration,
      setup: stateManager.setBroadcaster.bind(stateManager)
    }
  };
}

/**
 * export all helper methods that need an Node.js environment
 */
export { getExtProps, activate };
export * from './types';
