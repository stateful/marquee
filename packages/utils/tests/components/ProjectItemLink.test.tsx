import React from 'react'
import userEvent from '@testing-library/user-event'
import { render, screen } from '@testing-library/react'

import { ProjectItemLink, MarqueeWindow } from '../../src'

declare const window: MarqueeWindow

test('should render nothing if path is not provided', () => {
  const item = { id: 'foobar', body: 'foobar', workspaceId: null }
  const { container } = render(<ProjectItemLink item={item} />)
  expect(container).toBeEmptyDOMElement()
})

test('with path as icon only', async () => {
  window.vscode = { postMessage: jest.fn() } as any
  const item = { id: 'foobar', body: 'foobar', workspaceId: null, path: '/foo/bar:42' }
  render(<ProjectItemLink item={item} iconOnly={true} />)
  expect(screen.queryByText('bar:42')).not.toBeInTheDocument()
  await userEvent.click(screen.getByTestId('LinkIcon'))
  expect((window.vscode.postMessage as jest.Mock).mock.calls).toMatchSnapshot()
})

test('with path with label', async () => {
  window.vscode = { postMessage: jest.fn() } as any
  const item = { id: 'foobar', body: 'foobar', workspaceId: null, path: '/foo/bar:42' }
  render(<ProjectItemLink item={item} />)
  expect(screen.getByText('bar:43')).toBeInTheDocument()
})

test('without path and no git information', async () => {
  const item = { id: 'foobar', body: 'foobar', workspaceId: null, origin: '/foo/bar:42' }
  const { container } = render(<ProjectItemLink item={item} />)
  expect(container).toBeEmptyDOMElement()
})

test('without path but with git information with label', async () => {
  const item = {
    id: 'foobar',
    body: 'foobar',
    workspaceId: null,
    origin: '/foo/bar:42',
    commit: 'sha123',
    branch: 'main',
    gitUri: 'git@github.com:stateful/marquee#main'
  }
  render(<ProjectItemLink item={item} />)
  expect(screen.getByText('bar:43')).toBeInTheDocument()
  expect((screen.getByLabelText('Project Item Link') as HTMLAnchorElement).href)
    .toBe('https://github.com/stateful/marquee/blob/sha123/foo/bar#L43')
})

test('without path but with git information icon only', async () => {
  const item = {
    id: 'foobar',
    body: 'foobar',
    workspaceId: null,
    origin: '/foo/bar:42',
    commit: 'sha123',
    branch: 'main',
    gitUri: 'git@github.com:stateful/marquee#main'
  }
  render(<ProjectItemLink item={item} iconOnly={true} />)
  expect(screen.queryByText('bar:43')).not.toBeInTheDocument()
})
