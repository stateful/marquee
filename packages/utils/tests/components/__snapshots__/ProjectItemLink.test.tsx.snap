// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with path as icon only 1`] = `
Array [
  Array [
    Object {
      "west": Object {
        "execCommands": Array [
          Object {
            "args": Array [
              Object {
                "item": Object {
                  "body": "foobar",
                  "id": "foobar",
                  "path": "/foo/bar:42",
                  "workspaceId": null,
                },
              },
            ],
            "command": "marquee.link",
          },
        ],
      },
    },
  ],
]
`;
